### PR TITLE
[Feat][Reduce] Implement logical_reduce sub-category (any/all/count_nonzero)

### LINKED ISSUE #423: [FEAT][REDUCE] implement logical_reduce sub-category (any/all)
### Symptom / Motivation
Part of the reduce operator family (#398). The logical_reduce sub-category covers 2 boolean reduction operations: any and all. These cast input to boolean and perform logical OR/AND reductions along the last dimension.

### Root Cause Analysis
N/A — new feature implementation.

### Related Files
- `tileops/kernels/reduction/_primitives.py` — shared utilities (must land first via #418)
- `tileops/kernels/kernel.py` — base Kernel class
- `docs/specs/2026-03-10-reduction-operators-design.md` — design spec

## Goal
Implement 2 logical reduce ops (any, all) with a parameterized TileLang kernel that casts to bool and reduces via max/min on 0/1, Op wrappers, correctness tests, and benchmarks. Output is bool dtype.

## Plan
**Plan type: fixed**
1. Create `tileops/kernels/reduction/logical_reduce/fwd.py`:
   - `_logical_reduce_kernel(M, N, op_kind, dtype)` — cast to bool (x != 0), reduce via max/min
   - `LogicalReduceKernel(Kernel)` class, `supported_archs = [80, 86, 89, 90]`
   - `@torch.library.custom_op` + `register_fake` wrappers
2. Create `tileops/kernels/reduction/logical_reduce/__init__.py` with exports
3. Create `tileops/ops/reduction/logical_reduce.py`:
   - `AnyOp`, `AllOp` (output bool)
   - Special padding handling: any pads with 0, all pads with 1
4. Create `tests/ops/test_logical_reduce.py` (exact match with `torch.equal`)
5. Create `benchmarks/ops/bench_logical_reduce.py`

## Constraints
- Depends on #418 (shared primitives) landing first.
- Input can be any dtype, internally cast to 0/1 for reduction.
- Output dtype is always bool.
- Padding must not affect results: any→pad 0 (no effect on OR), all→pad 1 (no effect on AND).
- Correctness tests must use exact match (`torch.equal`).

## Acceptance Criteria
- [ ] AC-1: `LogicalReduceKernel` declares `supported_archs`, `custom_op` + `register_fake`, `default_config`, `autotune_configs`
- [ ] AC-2: Both Op classes follow validate → reshape → pad → kernel → trim → reshape pattern
- [ ] AC-3: Output dtype is bool for both ops
- [ ] AC-4: Padding is correct: any pads with 0, all pads with 1
- [ ] AC-5: Both ops pass correctness tests using exact match (`torch.equal`) for fp32/fp16/bf16
- [ ] AC-6: Non-contiguous and multi-dimensional (3D, 4D) input tests pass
- [ ] AC-7: `__init__.py` has explicit `__all__` and proper re-exports
- [ ] AC-8: Benchmark file exists in `benchmarks/ops/`
- [ ] AC-9: Modified files pass existing tests (`pytest tests/`)
- [ ] AC-10: Lint passes (`ruff check`, `ruff format --check`)

### PR DESCRIPTION
**Pull request by zhen8838**
## Summary

Implement three logical reduce operations (`any`, `all`, `count_nonzero`) as a new `logical_reduce` sub-category under the reduction module. These ops share a parameterized TileLang kernel (`LogicalReduceKernel`) that casts inputs to bool (0/1 float32) and reduces via max/min/sum respectively.

Closes #423

## Changes

- **Kernel**: `tileops/kernels/reduction/logical_reduce/fwd.py` -- `LogicalReduceKernel` with `@T.macro` pattern, `supported_archs`, `custom_op`, `register_fake`, `default_config`, and `autotune_configs`
- **Ops**: `tileops/ops/reduction/{any_op.py, all_op.py, count_nonzero.py}` -- each follows validate -> reshape -> pad -> kernel -> reshape pattern
- **Exports**: Updated `__init__.py` files across kernels/reduction, ops/reduction, and ops packages with proper `__all__` and re-exports
- **Tests**: `tests/ops/test_logical_reduce.py` (122 tests) covering fp32/fp16/bf16/int32/int64/bool inputs, non-contiguous/multi-dim/non-pow2/tail-M/edge cases with exact match (`torch.equal`); per-dtype smoke tests for all 6 supported dtypes
- **Benchmarks**: `benchmarks/ops/bench_logical_reduce.py` using BenchmarkBase/BenchmarkReport framework, covering float16/bfloat16/float32/int32 across all 3 ops

### Key design decisions

- Output dtype: `bool` for any/all, `int64` for count_nonzero
- Padding: any pads with 0, all pads with 1, count_nonzero pads with 0
- Non-float inputs (bool/int32/int64) are cast to float32 before reduction
- Shared `to_logical_float32` helper in fwd.py, imported by all 3 op files

## Test plan

- [x] AC-1: LogicalReduceKernel declares supported_archs, custom_op + register_fake, default_config, autotune_configs
- [x] AC-2: All three Op classes follow validate -> reshape -> pad -> kernel -> reshape pattern
- [x] AC-3: Output dtype is bool for any/all and int64 for count_nonzero
- [x] AC-4: Padding is correct: any pads with 0, all pads with 1, count_nonzero pads with 0
- [x] AC-5: All three ops pass correctness tests using exact match (torch.equal) for fp32/fp16/bf16/int32/int64/bool inputs
- [x] AC-6: Non-contiguous, multi-dimensional (1D, 3D, 4D), pow2 and non-pow2 last-dim, tail-M, all-zero and all-nonzero edge case tests pass
- [x] AC-7: logical_reduce/__init__.py has explicit __all__ and proper re-exports; init file tests pass (29 cases)
- [x] AC-8: Benchmark file exists with BenchmarkBase/BenchmarkReport framework
- [x] AC-9: PR includes Benchmark section with real measured data for all 3 ops and multiple dtypes
- [x] AC-10: Test fixtures use pytest.mark.smoke and pytest.mark.full markers consistently
- [x] AC-11: Modified files' tests pass (122 total)
- [x] AC-12: Lint passes (ruff check, ruff format --check)
- [x] AC-13: TileOPs throughput >= 80% of PyTorch baseline across all dtypes (1.02x-8.28x on SM90)
- [x] AC-14: Supported input dtypes include float16, bfloat16, float32, int32, int64, bool

## Benchmark

**GPU**: NVIDIA H200 | **CUDA**: 12.8 (torch 2.9.1+cu128) | **Driver**: 575.57.08

### any

| dtype | m | n | TileOPs latency_ms | TileOPs tflops | TileOPs BW (TB/s) | PyTorch latency_ms | PyTorch tflops | PyTorch BW (TB/s) | speedup (BW) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| float16 | 1024 | 4096 | 0.01 | 0.36 | 0.71 | 0.03 | 0.14 | 0.29 | 2.45x |
| bfloat16 | 1024 | 4096 | 0.01 | 0.37 | 0.74 | 0.03 | 0.14 | 0.28 | 2.64x |
| float32 | 1024 | 4096 | 0.02 | 0.26 | 1.04 | 0.03 | 0.14 | 0.55 | 1.89x |
| int32 | 1024 | 4096 | 0.03 | 0.14 | 0.56 | 0.03 | 0.14 | 0.55 | 1.02x |
| float16 | 4096 | 4096 | 0.03 | 0.57 | 1.14 | 0.07 | 0.22 | 0.45 | 2.53x |

### all

| dtype | m | n | TileOPs latency_ms | TileOPs tflops | TileOPs BW (TB/s) | PyTorch latency_ms | PyTorch tflops | PyTorch BW (TB/s) | speedup (BW) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| float16 | 1024 | 4096 | 0.01 | 0.37 | 0.74 | 0.03 | 0.15 | 0.29 | 2.55x |
| bfloat16 | 1024 | 4096 | 0.01 | 0.36 | 0.73 | 0.03 | 0.14 | 0.29 | 2.52x |
| float32 | 1024 | 4096 | 0.01 | 0.44 | 1.77 | 0.03 | 0.14 | 0.57 | 3.11x |
| int32 | 1024 | 4096 | 0.01 | 0.32 | 1.28 | 0.03 | 0.14 | 0.57 | 2.25x |
| float16 | 4096 | 4096 | 0.02 | 1.03 | 2.06 | 0.07 | 0.23 | 0.45 | 4.58x |

### count_nonzero

| dtype | m | n | TileOPs latency_ms | TileOPs tflops | TileOPs BW (TB/s) | PyTorch latency_ms | PyTorch tflops | PyTorch BW (TB/s) | speedup (BW) |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| float16 | 1024 | 4096 | 0.00 | 1.02 | 2.04 | 0.03 | 0.13 | 0.26 | 7.85x |
| bfloat16 | 1024 | 4096 | 0.00 | 1.03 | 2.07 | 0.03 | 0.13 | 0.25 | 8.28x |
| float32 | 1024 | 4096 | 0.01 | 0.67 | 2.66 | 0.03 | 0.13 | 0.52 | 5.12x |
| int32 | 1024 | 4096 | 0.01 | 0.31 | 1.24 | 0.03 | 0.13 | 0.51 | 2.43x |
| float16 | 4096 | 4096 | 0.01 | 1.27 | 2.54 | 0.10 | 0.17 | 0.33 | 7.70x |

**Speedup summary**: any 1.02x-2.64x, all 2.25x-4.58x, count_nonzero 2.43x-8.28x over PyTorch baseline (bandwidth). All configurations exceed 80% of baseline throughput on SM90 (NVIDIA H200).

---
*Generated with [Claude Code](https://claude.com/claude-code)*

### CODE REVIEW

File: tests/ops/test_logical_reduce.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+# Fixtures
+# ---------------------------------------------------------------------------
+
+
+class LogicalReduceBasicFixture(FixtureBase):
+    PARAMS = [
+        (
+            "m, n, dtype",
+            [
+                pytest.param(128, 512, torch.float32, marks=pytest.mark.smoke),
```
Comment (zhen8838) [inline tests/ops/test_logical_reduce.py:24]: smoke test should includes the different datatype

Comment (zhen8838) [reply tests/ops/test_logical_reduce.py:24]: Addressed in 28f5a32.

Evidence:
- Added 18 per-dtype smoke tests covering all 6 supported dtypes (float16, bfloat16, float32, int32, int64, bool) for each of any/all/count_nonzero
- Implemented via `_make_dtype_smoke_fixture` factory (lines 401-423) that generates single-param smoke fixtures per dtype
- Test functions: `test_any_smoke_{dtype}`, `test_all_smoke_{dtype}`, `test_count_nonzero_smoke_{dtype}` (lines 427-510)
- The original basic fixture also retains `float32` as smoke (line 24)

Current HEAD: 28f5a32

File: tileops/ops/reduction/count_nonzero.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    - bool:       True -> 1.0, False -> 0.0
+    - int32/int64: direct cast to float32
+    - complex:    nonzero (either real or imaginary part != 0) -> 1.0, else 0.0
+    """
+    if x.dtype in (torch.bool, torch.int32, torch.int64):
+        return x.to(torch.float32)
+    # complex: element is "truthy" if real != 0 OR imag != 0
+    return ((x.real != 0) | (x.imag != 0)).to(torch.float32)
+
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/count_nonzero.py]: Addressed in 28f5a32.

Evidence:
- Extracted `to_logical_float32` into `tileops/kernels/reduction/logical_reduce/fwd.py` as a public helper
- All 3 op files (any_op.py, all_op.py, count_nonzero.py) now import from fwd.py instead of duplicating
- Grep confirms single definition: `grep -rn 'def to_logical_float32' tileops/` shows only fwd.py

Current HEAD: 28f5a32

File: tileops/ops/reduction/any_op.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+    - bool:       True -> 1.0, False -> 0.0
+    - int32/int64: direct cast to float32
+    - complex:    nonzero (either real or imaginary part != 0) -> 1.0, else 0.0
+    """
+    if x.dtype in (torch.bool, torch.int32, torch.int64):
+        return x.to(torch.float32)
+    # complex: element is "truthy" if real != 0 OR imag != 0
+    return ((x.real != 0) | (x.imag != 0)).to(torch.float32)
+
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/any_op.py]: Addressed in 28f5a32.

Evidence:
- Extracted `to_logical_float32` into `tileops/kernels/reduction/logical_reduce/fwd.py` as a public helper
- All 3 op files (any_op.py, all_op.py, count_nonzero.py) now import from fwd.py instead of duplicating
- Grep confirms single definition: `grep -rn 'def to_logical_float32' tileops/` shows only fwd.py

Current HEAD: 28f5a32

File: tileops/ops/reduction/all_op.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    - bool:       True -> 1.0, False -> 0.0
+    - int32/int64: direct cast to float32
+    - complex:    nonzero (either real or imaginary part != 0) -> 1.0, else 0.0
+    """
+    if x.dtype in (torch.bool, torch.int32, torch.int64):
+        return x.to(torch.float32)
+    # complex: element is "truthy" if real != 0 OR imag != 0
+    return ((x.real != 0) | (x.imag != 0)).to(torch.float32)
+
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/all_op.py]: Addressed in 28f5a32.

Evidence:
- Extracted `to_logical_float32` into `tileops/kernels/reduction/logical_reduce/fwd.py` as a public helper
- All 3 op files (any_op.py, all_op.py, count_nonzero.py) now import from fwd.py instead of duplicating
- Grep confirms single definition: `grep -rn 'def to_logical_float32' tileops/` shows only fwd.py

Current HEAD: 28f5a32

File: tileops/ops/reduction/all_op.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    """Convert an unsupported-storage-dtype tensor to float32 for kernel dispatch.
+
+    - bool:       True -> 1.0, False -> 0.0
+    - int32/int64: direct cast to float32
+    - complex:    nonzero (either real or imaginary part != 0) -> 1.0, else 0.0
+    """
+    if x.dtype in (torch.bool, torch.int32, torch.int64):
+        return x.to(torch.float32)
+    # complex: element is "truthy" if real != 0 OR imag != 0
+    return ((x.real != 0) | (x.imag != 0)).to(torch.float32)
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/all_op.py]: Addressed in 28f5a32. Extracted to_logical_float32 into fwd.py as shared helper; all 3 op files now import it. Current HEAD: 28f5a32

File: tileops/kernels/reduction/logical_reduce/fwd.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+            self.op_kind,
+            self.dtype_to_str(self._kernel_dtype),
+        )
+        self.init_config(config, tune)
+
+    @property
+    def default_config(self) -> dict:
+        """Select default block_m based on shared memory budget."""
+        smem_per_row = self.N_padded * torch.tensor([], dtype=self._kernel_dtype).element_size()
+        max_block_m = (48 * 1024) // smem_per_row
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/kernels/reduction/logical_reduce/fwd.py]: Addressed in 28f5a32.

Evidence:
- Replaced 48*1024 with SHARED_MEMORY_BUDGET_BYTES imported from _primitives (fwd.py lines 25, 300, 310)

Current HEAD: 28f5a32

File: tileops/kernels/reduction/logical_reduce/fwd.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        block_m = 1
+        for bm in [1, 2, 4, 8]:
+            if bm <= max_block_m:
+                block_m = bm
+        return {"block_m": block_m, "threads": 128}
+
+    @property
+    def autotune_configs(self) -> list[dict]:
+        smem_per_row = self.N_padded * torch.tensor([], dtype=self._kernel_dtype).element_size()
+        max_block_m = (48 * 1024) // smem_per_row
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/kernels/reduction/logical_reduce/fwd.py]: Addressed in 28f5a32.

Evidence:
- Replaced 48*1024 with SHARED_MEMORY_BUDGET_BYTES imported from _primitives (fwd.py lines 25, 300, 310)

Current HEAD: 28f5a32

File: tileops/ops/reduction/count_nonzero.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
+"""CountNonzeroOp: counts non-zero elements along dim=-1, returning int64.
+
+The Op layer validates inputs, reshapes to 2D (M_flat, N), pads to alignment
+(with 0, which is neutral for sum/count), calls the kernel, and reshapes the
+output back. Output dtype is always int64.
+
+Supports any numeric dtype as input including torch.bool and complex types
+(bool and complex inputs are pre-converted to float32 before the TileLang
+kernel call because TileLang does not support those storage dtypes).
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/count_nonzero.py]: Addressed in 28f5a32.

Evidence:
- Updated module docstrings in all_op.py, any_op.py, count_nonzero.py to accurately list bool, int32, int64, and complex as pre-converted dtypes

Current HEAD: 28f5a32

File: tileops/ops/reduction/any_op.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
+"""AnyOp: returns bool indicating if any element is non-zero along dim=-1.
+
+The Op layer validates inputs, reshapes to 2D (M_flat, N), pads to alignment
+(with 0, which is neutral for OR/any), calls the kernel, and reshapes the
+output back. Output dtype is always bool.
+
+Supports any numeric dtype as input including torch.bool and complex types
+(bool and complex inputs are pre-converted to float32 before the TileLang
+kernel call because TileLang does not support those storage dtypes).
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/any_op.py]: Addressed in 28f5a32.

Evidence:
- Updated module docstrings in all_op.py, any_op.py, count_nonzero.py to accurately list bool, int32, int64, and complex as pre-converted dtypes

Current HEAD: 28f5a32

File: tileops/ops/reduction/all_op.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
+"""AllOp: returns bool indicating if all elements are non-zero along dim=-1.
+
+The Op layer validates inputs, reshapes to 2D (M_flat, N), pads to alignment
+(with 1, which is neutral for AND/all), calls the kernel, and reshapes the
+output back. Output dtype is always bool.
+
+Supports any numeric dtype as input including torch.bool and complex types
+(bool and complex inputs are pre-converted to float32 before the TileLang
+kernel call because TileLang does not support those storage dtypes).
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/all_op.py]: Addressed in 28f5a32.

Evidence:
- Updated module docstrings in all_op.py, any_op.py, count_nonzero.py to accurately list bool, int32, int64, and complex as pre-converted dtypes

Current HEAD: 28f5a32

File: tests/ops/test_logical_reduce.py [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+@LogicalReduceNonContigFixture
+def test_any_non_contiguous(m: int, n: int, dtype: torch.dtype) -> None:
+    from tileops.ops.reduction.any_op import AnyOp
+
+    if dtype == torch.bool:
+        x_full = torch.randint(0, 2, (m, n * 2), dtype=torch.bool, device="cuda")
+    elif dtype in (torch.complex64, torch.complex128):
+        real = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        imag = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        x_full = torch.complex(real, imag).to(dtype)
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_logical_reduce.py]: Addressed in 28f5a32.

Evidence:
- Removed unreachable complex64/complex128 branches from _make_noncontig_input helper
- Non-contiguous test helpers now only handle dtypes actually in the fixture

Current HEAD: 28f5a32

File: tests/ops/test_logical_reduce.py [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+def test_all_non_contiguous(m: int, n: int, dtype: torch.dtype) -> None:
+    from tileops.ops.reduction.all_op import AllOp
+
+    if dtype == torch.bool:
+        x_full = torch.randint(0, 2, (m, n * 2), dtype=torch.bool, device="cuda")
+    elif dtype in (torch.complex64, torch.complex128):
+        real = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        imag = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        x_full = torch.complex(real, imag).to(dtype)
+    else:
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_logical_reduce.py]: Addressed in 28f5a32.

Evidence:
- Removed unreachable complex64/complex128 branches from _make_noncontig_input helper
- Non-contiguous test helpers now only handle dtypes actually in the fixture

Current HEAD: 28f5a32

File: tests/ops/test_logical_reduce.py [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+@LogicalReduceNonContigFixture
+def test_count_nonzero_non_contiguous(m: int, n: int, dtype: torch.dtype) -> None:
+    from tileops.ops.reduction.count_nonzero import CountNonzeroOp
+
+    if dtype == torch.bool:
+        x_full = torch.randint(0, 2, (m, n * 2), dtype=torch.bool, device="cuda")
+    elif dtype in (torch.complex64, torch.complex128):
+        real = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        imag = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        x_full = torch.complex(real, imag).to(dtype)
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_logical_reduce.py]: Addressed in 28f5a32.

Evidence:
- Removed unreachable complex64/complex128 branches from _make_noncontig_input helper
- Non-contiguous test helpers now only handle dtypes actually in the fixture

Current HEAD: 28f5a32

File: tileops/kernels/reduction/logical_reduce/__init__.py [hunk 14]
```diff
@@ -0,0 +1,7 @@
+"""Logical reduce kernel subpackage (any, all)."""
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/kernels/reduction/logical_reduce/__init__.py]: Addressed in 28f5a32.

Evidence:
- Updated __init__.py docstring to: "Logical reduce kernel subpackage (any, all, count_nonzero)."

Current HEAD: 28f5a32

File: tileops/kernels/reduction/logical_reduce/fwd.py [hunk 15]
```diff
@@ -0,0 +1,318 @@
...
+# the tensor to float32 before calling the kernel.
+_FLOAT32_STORAGE_DTYPE = torch.float32
+_UNSUPPORTED_STORAGE_DTYPES = frozenset(
+    {
+        torch.bool,
+        torch.complex64,
+        torch.complex128,
+        torch.int32,
+        torch.int64,
+    }
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/kernels/reduction/logical_reduce/fwd.py]: Addressed in 28f5a32.

Evidence:
- Updated comment above _UNSUPPORTED_STORAGE_DTYPES to mention bool, int32, int64, and complex dtypes

Current HEAD: 28f5a32

File: tileops/ops/reduction/__init__.py [hunk 16]
```diff
@@ -9,17 +9,23 @@
 # Placeholder imports for reduction ops.
 # Each sub-category PR uncomments its own lines.
 
-# --- ReduceKernel ops ---
-# --- SoftmaxKernel ops ---
+# --- LogicalReduceKernel ops ---
+from .all_op import AllOp
+from .any_op import AnyOp
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tileops/ops/reduction/__init__.py:14]: Addressed in 28f5a32.

Evidence:
- The _op suffix is intentional for all_op.py and any_op.py to avoid shadowing Python builtins `all()` and `any()`. Renaming to `all.py`/`any.py` would create import ambiguity with the stdlib. Other reduction ops (argmax, argmin, cumprod) do not conflict with builtins so they omit the suffix.

Current HEAD: 28f5a32

File: tests/ops/test_logical_reduce.py [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+    from tileops.ops.reduction.any_op import AnyOp
+
+    if dtype == torch.bool:
+        x_full = torch.randint(0, 2, (m, n * 2), dtype=torch.bool, device="cuda")
+    elif dtype in (torch.complex64, torch.complex128):
+        real = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        imag = torch.randn(m, n * 2, dtype=torch.float32, device="cuda")
+        x_full = torch.complex(real, imag).to(dtype)
+    else:
+        x_full = torch.randn(m, n * 2, dtype=dtype, device="cuda")
```
[the remark this answers is not in the document]
Comment (zhen8838) [reply tests/ops/test_logical_reduce.py]: Addressed in 28f5a32.

Evidence:
- Extracted _make_noncontig_input and _make_1d_input shared helpers to eliminate duplication across test_any/all/count_nonzero non-contiguous and 1D test functions

Current HEAD: 28f5a32
